contracts/asset-proxy/contracts/src/bridges/DydxBridge.sol

```diff
@@ -0,0 +1,179 @@
+    /// @param amount Minimum amount of `toTokenAddress` tokens to buy.
+    /// @param encodedBridgeData An abi-encoded `BridgeData` struct.
+    /// @return success The magic bytes if successful.
+    function bridgeTransferFrom(
+        address,
+        address from,
+        address to,
+        uint256 amount,
+        bytes calldata encodedBridgeData
+    )
+        external
+        onlyAuthorized
+        returns (bytes4 success)
+    {
+        // Decode bridge data.
+        (BridgeData memory bridgeData) = abi.decode(encodedBridgeData, (BridgeData));
+
+        // Cache dydx contract.
+        IDydx dydx = IDydx(_getDydxAddress());
+
+        // Assert that `from` is the owner or an operator of the dydx account.
+        require(
+            from == bridgeData.accountOwner || dydx.getIsLocalOperator(bridgeData.accountOwner, from),
+            "INVALID_DYDX_OWNER_OR_OPERATOR"
+        );
+
+        // Construct dydx account info.
+        IDydx.AccountInfo[] memory accounts = new IDydx.AccountInfo[](1);
+        accounts[0] = IDydx.AccountInfo({
+            owner: bridgeData.accountOwner,
+            number: bridgeData.accountNumber
+        });
+
+        // Create dydx action.
+        IDydx.ActionArgs[] memory actions = new IDydx.ActionArgs[](1);
+        if (bridgeData.action == BridgeAction.Deposit) {
+            actions[0] = _createDepositAction(
+                from,
+                amount,
+                bridgeData
+            );
+        } else if (bridgeData.action == BridgeAction.Withdraw) {
+            actions[0] = _createWithdrawAction(
+                to,
+                amount,
+                bridgeData
+            );
+        } else {
+            // If all values in the `Action` enum are handled then this
+            // revert is unreachable: Solidity will revert when casting
+            // from `uint8` to `Action`.
+            revert("UNRECOGNIZED_ACTION");
+        }
+
+        // Run operation. This will revert on failure.
+        dydx.operate(accounts, actions);
+        return BRIDGE_SUCCESS;
+    }
+
+    /// @dev Returns a dydx `DepositAction`.
+    /// @param depositFrom Deposit tokens from this address.
+    /// @param amount of tokens to deposit.
+    /// @param bridgeData A `BridgeData` struct.
+    function _createDepositAction(
+        address depositFrom,
+        uint256 amount,
+        BridgeData memory bridgeData
+    )
+        internal
+        pure
+        returns (IDydx.ActionArgs memory)
+    {
+        // Create dydx amount.
+        IDydx.AssetAmount memory amountToDeposit = IDydx.AssetAmount({
+            sign: true,                                 // true if positive.
+            denomination: IDydx.AssetDenomination.Wei,  // Wei => actual token amount held in account.
+            ref: IDydx.AssetReference.Target,           // Target => an absolute amount.
+            value: amount                               // amount to deposit.
```

Comment:
This introduces potential rounding errors because the `amount` parameter has already been rounded down in the exchange. Digging into this, it would have been introduced by the MAP version as well.

```
## In Exchange
The `amount` parameter is computed by the Exchange as:
  amount = floor[(takerAssetFillAmount * makerAssetAmount) / takerAssetAmount]

The amount to deposit is `takerAssetFillAmount`, which could be recomputed in the Bridge as:
  takerAssetFillAmount = (amount * takerAssetAmount) / makerAssetAmount

However since `amount` is rounded we cannot precisely extract the original `takerAssetAmount`:
  {
      makerAssetAmount: 3,
      takerAssetAmount: 11,
      takerAssetFillAmount: 5
  }

  In Exchange:
  amount = floor[(5 * 3) / 11] = floor[1.3636..] = 1

  In Bridge:
  amountToDeposit = takerAssetFillAmount = (1 * 11) / 3 = 3.6666 (neither rounding up nor down would give 5)
```

If we split the operations between the maker and taker asset data.
Doing this we transfer the exact values computed by the Exchange:
```
Order: {
  ...,
  takerAssetData: [ERC20 token to Maker, Bridge deposits from Maker (Maker=`to`)]
  makerAssetData: [Bridge withdraws to Taker (Maker=`from`)]
}
```